### PR TITLE
Feature/short circuit main model

### PR DESCRIPTION
**Pull request by mgovers**
Implement the main model for short circuit.

NOTE: this PR is based off both

* [x] #307 ; and
* [x] #308 

Do not merge until both are merged

### CODE REVIEW

**Comment by mgovers** [comment]
There is still an issue with the topology but i was able to make the validation tests run (although they give segfaults at this point). My proposal is to still review + merge this PR and then fix the validation tests in a separate PR to test the full workflow of this feature.

In the meantime, I explicitly marked the `short-circuit` as unsupported in the C API

**Comment by TonyXiang8787** [comment]
@mgovers can you fix the CI errors first?

**Comment by mgovers** [comment]
Coverage is low due to the validation tests not yet running

File: power_grid_model_c/power_grid_model/include/power_grid_model/main_model.hpp [hunk 1]
```diff
@@ -1003,6 +1109,53 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
         return se_input;
     }
 
+    template <bool sym>
+    std::vector<ShortCircuitInput> prepare_short_circuit_input() {
+        assert(is_topology_up_to_date_ && is_parameter_up_to_date<sym>());
+
+        std::vector<ShortCircuitInput> sc_input(n_math_solvers_);
+        for (auto& solver_input : sc_input) {
+            solver_input.fault_bus_indptr = {0};
```
Comment (TonyXiang8787) [inline power_grid_model_c/power_grid_model/include/power_grid_model/main_model.hpp:1118]: @mgovers  we have a separate function just to sort and generate any `indptr` like structure. Please use that:

https://github.com/PowerGridModel/power-grid-model/blob/feature/short-circuit-main-model/power_grid_model_c/power_grid_model/include/power_grid_model/sparse_mapping.hpp

**Comment by TonyXiang8787** [comment]
> FYI: I believe I have fixed the topology bug. I tried running validation tests again but I got sparse matrix errors (see below). I would suggest keeping that out of scope of this PR, because there's no guarantee that the issue lies in the main model. The C API and Python API are unaffected for the time being, because short circuit calculations still are explicitly unsupported by the API.
> 
> ```
> ===============================================================================
> 62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(605):
> 62: TEST CASE:  Validation test batch
> 62:   short_circuit/single_phase_to_ground-asym-iec60909_batch
> 62: 
> 62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(614): ERROR: Unexpected exception with message: Sparse matrix error, possibly singular matrix!
> 62: If you get this error from state estimation, it usually means the system is not fully observable, i.e. not enough measurements.
> 62: 
> 62: Validation test: short_circuit/three_phase-asym-iec60909_batch
> 62: ===============================================================================
> 62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(605):
> 62: TEST CASE:  Validation test batch
> 62:   short_circuit/three_phase-asym-iec60909_batch
> 62: 
> 62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(614): ERROR: Unexpected exception with message: Sparse matrix error, possibly singular matrix!
> 62: If you get this error from state estimation, it usually means the system is not fully observable, i.e. not enough measurements.
> 62: 
> 62: Validation test: short_circuit/two_phase-asym-iec60909_batch
> 62: ===============================================================================
> 62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(605):
> 62: TEST CASE:  Validation test batch
> 62:   short_circuit/two_phase-asym-iec60909_batch
> 62: 
> 62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(614): ERROR: Unexpected exception with message: Sparse matrix error, possibly singular matrix!
> 62: If you get this error from state estimation, it usually means the system is not fully observable, i.e. not enough measurements.
> 62: 
> 62: Validation test: short_circuit/two_phase_to_ground-asym-iec60909_batch
> 62: ===============================================================================
> 62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(605):
> 62: TEST CASE:  Validation test batch
> 62:   short_circuit/two_phase_to_ground-asym-iec60909_batch
> 62: 
> 62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(614): ERROR: Unexpected exception with message: Sparse matrix error, possibly singular matrix!
> 62: If you get this error from state estimation, it usually means the system is not fully observable, i.e. not enough measurements.
> 62: 
> ```

Agree to solve the validation test issue in a separate PR and then enable C and Python API. Please in this PR use the generic sparse mapping technique to generate `fault_bus_indptr`.

**Comment by mgovers** [comment]
I did but then i got the segfaults i had earlier today. Will revert tomorrow and see if the issue just resolved itself but i was unable to find what the issue was exactly. Also, most things actually are private within the topology class so that complicates things as well

Op 26 jul. 2023 20:52 schreef Tony Xiang ***@***.***>:

FYI: I believe I have fixed the topology bug. I tried running validation tests again but I got sparse matrix errors (see below). I would suggest keeping that out of scope of this PR, because there's no guarantee that the issue lies in the main model. The C API and Python API are unaffected for the time being, because short circuit calculations still are explicitly unsupported by the API.

===============================================================================
62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(605):
62: TEST CASE:  Validation test batch
62:   short_circuit/single_phase_to_ground-asym-iec60909_batch
62:
62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(614): ERROR: Unexpected exception with message: Sparse matrix error, possibly singular matrix!
62: If you get this error from state estimation, it usually means the system is not fully observable, i.e. not enough measurements.
62:
62: Validation test: short_circuit/three_phase-asym-iec60909_batch
62: ===============================================================================
62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(605):
62: TEST CASE:  Validation test batch
62:   short_circuit/three_phase-asym-iec60909_batch
62:
62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(614): ERROR: Unexpected exception with message: Sparse matrix error, possibly singular matrix!
62: If you get this error from state estimation, it usually means the system is not fully observable, i.e. not enough measurements.
62:
62: Validation test: short_circuit/two_phase-asym-iec60909_batch
62: ===============================================================================
62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(605):
62: TEST CASE:  Validation test batch
62:   short_circuit/two_phase-asym-iec60909_batch
62:
62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(614): ERROR: Unexpected exception with message: Sparse matrix error, possibly singular matrix!
62: If you get this error from state estimation, it usually means the system is not fully observable, i.e. not enough measurements.
62:
62: Validation test: short_circuit/two_phase_to_ground-asym-iec60909_batch
62: ===============================================================================
62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(605):
62: TEST CASE:  Validation test batch
62:   short_circuit/two_phase_to_ground-asym-iec60909_batch
62:
62: C:\Users\MartijnGovers\Documents\power-grid-model\tests\cpp_validation_tests\test_validation.cpp(614): ERROR: Unexpected exception with message: Sparse matrix error, possibly singular matrix!
62: If you get this error from state estimation, it usually means the system is not fully observable, i.e. not enough measurements.
62:


Agree to solve the validation test issue in a separate PR and then enable C and Python API. Please in this PR use the generic sparse mapping technique to generate fault_bus_indptr.

File: power_grid_model_c/power_grid_model/include/power_grid_model/main_model.hpp [hunk 2]
```diff
@@ -1003,6 +1109,51 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
...
 
+    template <bool sym>
+    std::vector<ShortCircuitInput> prepare_short_circuit_input() {
+        assert(is_topology_up_to_date_ && is_parameter_up_to_date<sym>());
+
+        std::vector<IdxVector> topo_fault_indices(state_.math_topology.size());
+        std::vector<IdxVector> topo_bus_indices(state_.math_topology.size());
+
+        for (Idx fault_idx{0}; fault_idx < state_.components.template size<Fault>(); ++fault_idx) {
+            auto const& fault = state_.components.template get_item_by_seq<Fault>(fault_idx);
```
Comment (TonyXiang8787) [inline power_grid_model_c/power_grid_model/include/power_grid_model/main_model.hpp:1120]: we only consider the faults which has `status()` returns `true`.  So we need a check here. If the `status` is `false`, we `continue`.

**Review by TonyXiang8787 — APPROVED** [review]

**Pull request merged**